org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/popup/AbstractNotificationPopup.java

```diff
@@ -0,0 +1,567 @@
+
+    shell = newShell;
+    newShell.setBackground(color.getBorder());
+  }
+
+  @Override
+  public void create() {
+    super.create();
+    addRegion(shell);
+  }
+
+  private void addRegion(Shell shell) {
+    Region region = new Region();
+    Point s = shell.getSize();
+
+    /* Add entire Shell */
+    region.add(0, 0, s.x, s.y);
+
+    /* Subtract Top-Left Corner */
+    region.subtract(0, 0, 5, 1);
+    region.subtract(0, 1, 3, 1);
+    region.subtract(0, 2, 2, 1);
+    region.subtract(0, 3, 1, 1);
+    region.subtract(0, 4, 1, 1);
+
+    /* Subtract Top-Right Corner */
+    region.subtract(s.x - 5, 0, 5, 1);
+    region.subtract(s.x - 3, 1, 3, 1);
+    region.subtract(s.x - 2, 2, 2, 1);
+    region.subtract(s.x - 1, 3, 1, 1);
+    region.subtract(s.x - 1, 4, 1, 1);
+
+    /* Subtract Bottom-Left Corner */
+    region.subtract(0, s.y, 5, 1);
+    region.subtract(0, s.y - 1, 3, 1);
+    region.subtract(0, s.y - 2, 2, 1);
+    region.subtract(0, s.y - 3, 1, 1);
+    region.subtract(0, s.y - 4, 1, 1);
+
+    /* Subtract Bottom-Right Corner */
+    region.subtract(s.x - 5, s.y - 0, 5, 1);
+    region.subtract(s.x - 3, s.y - 1, 3, 1);
+    region.subtract(s.x - 2, s.y - 2, 2, 1);
+    region.subtract(s.x - 1, s.y - 3, 1, 1);
+    region.subtract(s.x - 1, s.y - 4, 1, 1);
+
+    /* Dispose old first */
+    if (shell.getRegion() != null) {
+      shell.getRegion().dispose();
+    }
+
+    /* Apply Region */
+    shell.setRegion(region);
+
+    /* Remember to dispose later */
+    lastUsedRegion = region;
+  }
+
+  private boolean isMouseOver(Shell shell) {
+    if (display.isDisposed()) {
+      return false;
+    }
+    return shell.getBounds().contains(display.getCursorLocation());
+  }
+
+  @Override
+  public int open() {
+    if (shell == null || shell.isDisposed()) {
+      shell = null;
+      create();
+    }
+
+    constrainShellSize();
+    shell.setLocation(fixupDisplayBounds(shell.getSize(), shell.getLocation()));
+
+    if (isFadingEnabled()) {
+      shell.setAlpha(0);
+    }
+    shell.setVisible(true);
+    fadeJob = AnimationUtil.fadeIn(shell, new IFadeListener() {
+      public void faded(Shell shell, int alpha) {
+        if (shell.isDisposed()) {
+          return;
+        }
+
+        if (alpha == 255) {
+          scheduleAutoClose();
+        }
+      }
+    });
+
+    return Window.OK;
+  }
+
+  protected void scheduleAutoClose() {
+    if (delayClose > 0) {
+      closeJob.schedule(delayClose);
+    }
+  }
+
+  @Override
+  protected Control createContents(Composite parent) {
+    ((GridLayout) parent.getLayout()).marginWidth = 1;
+    ((GridLayout) parent.getLayout()).marginHeight = 1;
+
+    /* Outer Composite holding the controls */
+    final Composite outerCircle = new Composite(parent, SWT.NO_FOCUS);
+    outerCircle.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
+    outerCircle.setBackgroundMode(SWT.INHERIT_FORCE);
+
+    outerCircle.addControlListener(new ControlAdapter() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this anonymous class number of lines from 62 to at most 20, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1188)